### PR TITLE
add dir/csv support to files & refactor

### PR DESCRIPTION
**Pull request by zramsay**
#157 
#158 
#159 
#160

### CODE REVIEW

File: ipfs/api_handler.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return []byte(""), err
+	}
+
+	var errs struct {
+		Message string
+		Code    int
+	}
+	if response.StatusCode >= http.StatusBadRequest {
```
Comment (compleatang) [inline ipfs/api_handler.go]: nice. didn't know could do that.

File: ipfs/readers.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"time"
+)
+
+//had to be here because import cycle :(
```
Comment (compleatang) [inline ipfs/readers.go]: i was like, why's this here....? then i saw the comment. :( indeed.

File: ipfs/url_handler.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		return fmt.Sprintf("%s%s", IPFSUrl(), ":8080/ipfs/")
+	}
+}
+
+func IPFSBaseAPIUrl() string {
+	return fmt.Sprintf("%s%s", IPFSUrl(), ":5001/api/v0/")
+}
+
+func sexyUrl() string {
+	//TODO load balancer (when one isn't enough)
```
Comment (compleatang) [inline ipfs/url_handler.go]: ye. we can handle that server side though.

**Comment by compleatang** [comment]
LGTM. Great work!

File: ipfs/url_handler.go (hunk 3 above)
Comment (zramsay) [inline ipfs/url_handler.go]: ya, just put it there as a reminder

**Pull request merged**